flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java

```diff
@@ -310,8 +309,32 @@ private void flush() {
 
         inFlightRequestsCount++;
         submitRequestEntries(batch, requestResult);
-        numRecordsOutCounter.inc(batchSize);
+    }
+
+    /**
+     * Creates the next batch of request entries while respecting the {@code maxBatchSize} and
+     * {@code flushOnBufferSizeBytes}. Also adds these to the metrics counters.
+     */
+    private List<RequestEntryT> createNextAvailableBatch() {
+        int batchSize = Math.min(maxBatchSize, bufferedRequestEntries.size());
+        List<RequestEntryT> batch = new ArrayList<>(batchSize);
+
+        int batchSizeBytes = 0;
+        for (int i = 0; i < batchSize; i++) {
+            if (batchSizeBytes + bufferedRequestEntries.peek().getSize()
+                    >= maxBatchSizeInBytes) {
+                break;
+            }
+            RequestEntryWrapper<RequestEntryT> elem = bufferedRequestEntries.remove();
+            batch.add(elem.getRequestEntry());
+            bufferedRequestEntriesTotalSizeInBytes -= elem.getSize();
+            batchSizeBytes += elem.getSize();
```

Comment:
Calling `.getSize()` multiple in the loop. Is it worth caching this value to a variable? Is it an expensive call?

Reply:
It's retrieving a primitive long field from a class, so cheap to call.... but will refactor for readability.